README.md
configured-rabbitmq-client
==========================
A small wrapper around the foo-foo-mq to allow configuration from confit.

## Persistent messages
exchanges and queues are `durable` by default. You can make messages persistent in an exchangeGroup (primary, retry and rejected) by doing the following:
```
{
  "exchangeGroups": {
    "exchange.test.request.v1": {
      "persistent": true,
      "keys": "exchange.test"
    }
  }
}
```

 Follow [this link](https://github.com/gas-buddy/wiki/wiki/Persistent-rabbitMQ-messages) to see more options

## TTL
Only retry queue use TTL by default. You can set it by doing the following.

NOTE: **You cannot change TTL on an existing queue**.
```
{
  "exchangeGroups": {
    "exchange.test.request.v1": {
      ...
      "perMessageTtl": true, // It should be used on new exchange groups even if you don't change the default value
      "retryDelay": 60000, // default 10000ms
    }
  }
}
```


CHANGELOG.md
7.0.0
=====
* Set `replyQueue` to false by default, avoiding an unused queue
* If a subscription goes against a missing queue, try with '.q' appended in case the caller used an exchange name instead of a queue name
* Assert when subscribing to a non-existent queue, so we get a better error message

9.0.0
=====
* Switch to foo-foo-mq which is just rabbot with updated deps


tests/test_config.js
import tap from 'tap';
import net from 'net';
import bluebird from 'bluebird';
import RabbitMQClient from '../src/index';

const mqConfig = {
  hostname: process.env.RABBIT_HOST || 'rabbitmq',
  port: process.env.RABBIT_PORT || 5672,
  username: process.env.RABBIT_USER || 'guest',
  password: process.env.RABBIT_PASSWORD || 'guest',
  logging: {
    level: 2,
  },
};

function configWithExchangeGroups(exchangeGroups) {
  return {
    ...mqConfig,
    config: {
      exchangeGroups,
    },
  };
}

const ctx = { logger: console };

tap.test('wait for rabbit', async (t) => {
  for (let i = 0; i < 10; i += 1) {
    let connected = false;
    // eslint-disable-next-line no-await-in-loop
    await new Promise((accept) => {
      const s = new net.Socket();
      s.once('error', () => {
        if (!connected) {
          t.ok(true, `Waiting for RabbitMQ response from ${mqConfig.hostname}:${mqConfig.port}`);
          setTimeout(accept, 2500);
        }
      });
      s.connect({
        host: mqConfig.hostname,
        port: mqConfig.port,
      }, () => {
        connected = true;
        s.end();
        setTimeout(accept, 2500);
      });
    });
    if (connected) {
      t.ok(true, `RabbitMQ found on ${mqConfig.host}:${mqConfig.port}`);
      return;
    }
  }
  t.fail('Could not connect to RabbitMQ');
});

tap.test('test exchange group retry', async (t) => {
  const retryCount = 5;
  t.plan((retryCount + 3) + retryCount);
  const mq = new RabbitMQClient(ctx, configWithExchangeGroups({
    test: {
      retries: retryCount,
      retryDelay: 250,
      keys: 'testkey',
    },
  }));
  await mq.start(ctx);
  let counter = 0;
  const errorMessage = 'retry again';

  await new Promise(async (accept) => {
    await mq.subscribe('test', 'testkey',
      async (context, message) => {
        if (counter > 0) {
          t.equal(message.properties.headers.error, errorMessage, 'Previous error message is written to message headers');
        } else {
          t.equal(RabbitMQClient.activeMessages.size, 1, 'Should have 1 active message');
        }
        counter += 1;
        t.ok(true, `Recieved messsage for the ${counter} time.`);
        if (counter === retryCount + 1) {
          await message.ack();
          accept();
        }
        throw new Error(errorMessage);
      });
    await mq.publish('test', 'testkey', {});
  }).then(async () => {
    t.equal(RabbitMQClient.activeMessages.size, 0, 'Should have 0 active message');
    await mq.stop(ctx);
  });
});

tap.test('test routing key reuse', async (t) => {
  const mq = new RabbitMQClient(ctx, configWithExchangeGroups({
    one: {
      keys: 'number',
    },
    two: {
      keys: 'number',
    },
  }));

  await mq.start(ctx);
  let receivedOne = 0;
  let receivedTwo = 0;

  await (async () => {
    await mq.subscribe('one', 'number', async (context, message) => {
      receivedOne += 1;
      await message.ack();
    });

    await mq.subscribe('two', 'number', async (context, message) => {
      receivedTwo += 1;
      await message.ack();
    });

    await mq.publish('one', 'number', {});
    await mq.publish('two', 'number', {});
    await mq.publish('two', 'number', {});
    await bluebird.delay(250);
  })();

  t.strictEqual(receivedOne, 1, 'Should have received exactly one message on the one handler.');
  t.strictEqual(receivedTwo, 2, 'Should have received exactly two messages on the two handler.');
  await mq.stop(ctx);
});

tap.test('test delivery_mode pass thru', async (t) => {
  const retryCount = 2;
  const retryDelay = 380;
  const mq = new RabbitMQClient(ctx, configWithExchangeGroups({
    persistent: {
      retries: retryCount,
      retryDelay,
      persistent: true,
      perMessageTtl: true,
      keys: 'one',
    },
  }));
  await mq.start(ctx);
  let lastMsgDelivery = null;
  await new Promise(async (accept) => {
    await mq.subscribe('persistent', 'one',
      async (context, message) => {
        const deliveryTime = (new Date()).getTime();
        if (!message.properties.headers.retryCount) {
          t.strictEqual(message.properties.deliveryMode, 2, 'Persistent message should have deliveryMode=2');
        } else {
          const deliveryDelay = deliveryTime - lastMsgDelivery;
          t.ok(deliveryDelay >= retryDelay && deliveryDelay < retryDelay + 10, 'Retry delivery should respect retryDelay within a margin of 10ms');
          t.strictEqual(message.properties.deliveryMode, 2, 'Persistent message retries should retain deliveryMode=2');
        }
        lastMsgDelivery = deliveryTime;
        throw new Error('error');
      });

    const failTimeout = setTimeout(() => {
      t.ok(false, 'Should have a message in the rejected queue after 5s');
      accept();
    }, 5000);
    await mq.subscribe('persistent.rejected.q', 'one',
      async (context, message) => {
        t.strictEqual(message.properties.deliveryMode, 2, 'Rejected persistent message should retain deliveryMode=2');
        message.ack();
        clearTimeout(failTimeout);
        accept();
      });
    await mq.publish('persistent', 'one', {});
  });
  t.equal(RabbitMQClient.activeMessages.size, 0, 'Should have 0 active message');
  await mq.stop(ctx);
});


tests/test_connection.js
import tap from 'tap';
import net from 'net';
import RabbitMQClient from '../src/index';

const mqConfig = {
  hostname: process.env.RABBIT_HOST || 'rabbitmq',
  port: process.env.RABBIT_PORT || 5672,
  username: process.env.RABBIT_USER || 'guest',
  password: process.env.RABBIT_PASSWORD || 'guest',
};

const ctx = { logger: console };

tap.test('wait for rabbit', async (t) => {
  for (let i = 0; i < 10; i += 1) {
    let connected = false;
    // eslint-disable-next-line no-await-in-loop
    await new Promise((accept) => {
      const s = new net.Socket();
      s.once('error', () => {
        if (!connected) {
          t.ok(true, `Waiting for RabbitMQ response from ${mqConfig.hostname}:${mqConfig.port}`);
          setTimeout(accept, 2500);
        }
      });
      s.connect({
        host: mqConfig.hostname,
        port: mqConfig.port,
      }, () => {
        connected = true;
        s.end();
        setTimeout(accept, 2500);
      });
    });
    if (connected) {
      t.ok(true, `RabbitMQ found on ${mqConfig.host}:${mqConfig.port}`);
      return;
    }
  }
  t.fail('Could not connect to RabbitMQ');
});

tap.test('test_connection', async (t) => {
  const mq = new RabbitMQClient(ctx, mqConfig);
  const client = await mq.start(ctx);
  t.ok(client.publish, 'Should have a publish method');
  await mq.stop(ctx);
  t.ok(true, 'Should shut down');
});

tap.test('test_connection with array hostname', async (t) => {
  const config = { ...mqConfig, hostname: [mqConfig.hostname, mqConfig.hostname] };
  const mq = new RabbitMQClient(ctx, config);
  const client = await mq.start(ctx);
  t.ok(client.publish, 'Should have a publish method');
  await mq.stop(ctx);
  t.ok(true, 'Should shut down');
});


tests/test_exchange_groups.js
import tap from 'tap';
import _ from 'lodash';
import { normalizeExchangeGroups, fooFooConfigFromExchangeGroups } from '../src/exchangeGroups';

const simpleCase = {
  keys: 'simplekey',
};

const noRetryCase = {
  retries: 0,
  keys: 'noretrykey',
};

const complexCase = {
  retryDelay: 10,
  exchange: 'conf.ex',
  queue: {
    name: 'autodeletequeue',
    autoDelete: true,
  },
  retryQueue: 'crazyretryqueuename',
  keys: ['somekey', 'someotherkey'],
};

const rejectedOnlyCase = {
  retries: 0,
  rejectedExchange: true,
  keys: 'rejectedonlykey',
};

const perMessageTtl = {
  perMessageTtl: true,
  keys: 'somekey',
};

const exchangeGroups = {
  simpleCase,
  noRetryCase,
  complexCase,
  rejectedOnlyCase,
  perMessageTtl,
};


tap.test('test exchange group normalization', async (t) => {
  const finalGroups = normalizeExchangeGroups(exchangeGroups);

  t.ok(finalGroups.simpleCase.exchange.name === 'simpleCase', 'Exchange should be named the same as the group');
  t.ok(finalGroups.simpleCase.queue.name === 'simpleCase.q', 'Exchange should be named the same as the group with q');
  t.ok(finalGroups.simpleCase.retries > 0
       && finalGroups.simpleCase.retryDelay > 0
       && finalGroups.simpleCase.retryExchange
       && finalGroups.simpleCase.retryQueue
       && finalGroups.simpleCase.rejectedExchange
       && finalGroups.simpleCase.rejectedQueue, 'Retry queues should be generated by default');

  t.ok(!finalGroups.noRetryCase.retryExchange
       && !finalGroups.noRetryCase.retryQueue
       && !finalGroups.noRetryCase.rejectedExchange
       && !finalGroups.noRetryCase.rejectedQueue, 'Retry queues should not be generated when retries is set to 0');

  t.ok(finalGroups.complexCase.queue.name === complexCase.queue.name, 'Deep queue name is accepted');
  t.ok(finalGroups.complexCase.queue.autoDelete === complexCase.queue.autoDelete, 'Deep queue properties are accepted');
  t.ok(finalGroups.complexCase.retryQueue.name === complexCase.retryQueue, 'Additional queue overrides are accepted.');
  t.ok(finalGroups.complexCase.retryQueue.messageTtl === complexCase.retryDelay, 'messageTtl should match retryDelay on retry queue');

  t.ok(finalGroups.rejectedOnlyCase.exchange
       && finalGroups.rejectedOnlyCase.queue
       && !finalGroups.rejectedOnlyCase.retryExchange
       && !finalGroups.rejectedOnlyCase.retryQueue
       && finalGroups.rejectedOnlyCase.rejectedExchange
       && finalGroups.rejectedOnlyCase.rejectedQueue && 'Rejected queue should be generated when specified with zero retries');

  t.ok(finalGroups.perMessageTtl.retryQueue.messageTtl === undefined, 'retryQueue messageTtl should be undefined when perMessageTtl is set');
});

tap.test('test exchange group to rabbit config translation', async (t) => {
  const finalGroups = normalizeExchangeGroups(exchangeGroups);
  const config = fooFooConfigFromExchangeGroups(finalGroups);

  t.ok(_.find(config.queues, ['name', finalGroups.complexCase.queue.name]).autoDelete === complexCase.queue.autoDelete, 'Specific configuration propogates to rabbit config');
  t.ok(_.find(config.exchanges, ['name', finalGroups.rejectedOnlyCase.exchange.name]), 'No retry rejected exchange gets made');
});


src/util/repl.js
import repl from 'repl';
import path from 'path';
// eslint-disable-next-line import/no-extraneous-dependencies
import winston from 'winston';
import Client from '../';

// eslint-disable-next-line import/no-dynamic-require
const qConfig = require(path.resolve(process.argv[2]));

const context = {
  logger: winston,
};

class Subscription {
  constructor(name, type, mqClient, thunk) {
    this.thunk = thunk;
    this.name = name;
    this.type = type;
    this.mqClient = mqClient;
    this.messages = [];
  }

  unsubscribe() {
    this.mqClient.unsubscribe(this.thunk);
  }

  publish(message) {
    return this.mqClient.publish(this.name, this.type, message);
  }

  async ack(ix) {
    await this.messages[ix].message.ack();
    this.messages[ix].accept();
    this.messages[ix] = null;
  }

  async nack(ix) {
    await this.messages[ix].message.nack();
    this.messages[ix].accept();
    this.messages[ix] = null;
  }

  async reject(ix) {
    await this.messages[ix].message.reject();
    this.messages[ix].accept();
    this.messages[ix] = null;
  }

  message(ix) {
    return this.messages[ix].message;
  }

  throw(ix, error) {
    this.messages[ix].reject(error);
    this.messages[ix] = null;
  }

  onMessage(message) {
    let ix = this.messages.findIndex(i => !i);
    if (ix === -1) {
      ix = this.messages.length;
    }
    const rz = {
      message,
    };
    this.messages[ix] = rz;
    const promise = new Promise((accept, reject) => {
      rz.accept = accept;
      rz.reject = reject;
    });
    winston.info(`Received message #${ix} on ${this.name}#${this.type}
-----------------------------------------------------------------------------
${JSON.stringify(message, null, '\t')}
-----------------------------------------------------------------------------
`);
    return promise;
  }
}

const mqClient = new Client(context, qConfig);
mqClient.contextFunction = () => ({
  gb: {
    logger: winston,
    wrapError(e) { return e; },
  },
});

mqClient
  .start(context).then(() => {
    const rl = repl.start('> ');
    rl.on('exit', () => mqClient.stop());
    rl.context.client = mqClient;
    rl.context.subs = {};

    rl.context.subscribe = (queueName, type) => {
      let sub;
      const thunk = mqClient.subscribe(queueName, type, (mqctx, message) => sub.onMessage(message));
      sub = new Subscription(queueName, type, mqClient, thunk);
      rl.context.subs[queueName] = rl.context.subs[queueName] || {};
      rl.context.subs[queueName][type] = sub;
      return sub;
    };
    rl.context.async = promise => promise.then(winston.info).catch(winston.error);
  })
  .catch(error => winston.error(error));


src/WrappedMessage.js

const messagesInFlight = new Set();

const CALL_INFO = Symbol('Call Info');
const CLIENT_INFO = Symbol('RabbitMQ Client');

function messageComplete(wrapped, eventName) {
  try { messagesInFlight.delete(wrapped); } catch (_) { /* noop */ }
  if (wrapped[CLIENT_INFO].listenerCount(eventName)) {
    wrapped[CLIENT_INFO].emit(eventName, wrapped[CALL_INFO]);
  }
  wrapped[CLIENT_INFO].emit('count', messagesInFlight.size);
}

export class WrappedMessage {
  constructor(client, message) {
    const { nack, reject, ack, rabbitMessage, ...rest } = message;
    Object.assign(this, rest);
    this.arrivalTime = Date.now();
    this.rabbitMessage = message;
    messagesInFlight.add(this);
    this[CLIENT_INFO] = client;
    this[CALL_INFO] = {
      operationName: 'handleQueueMessage',
      message,
    };
    client.emit('start', this[CALL_INFO]);
    client.emit('count', messagesInFlight.size);
  }

  async ack() {
    try {
      await this.rabbitMessage.ack();
    } catch (error) {
      throw error;
    } finally {
      messageComplete(this, 'finish');
    }
  }

  async nack() {
    try {
      await this.rabbitMessage.nack();
    } catch (error) {
      throw error;
    } finally {
      messageComplete(this, 'error');
    }
  }

  async reject() {
    try {
      await this.rabbitMessage.reject();
    } catch (error) {
      throw error;
    } finally {
      messageComplete(this, 'error');
    }
  }

  static activeMessages() {
    return messagesInFlight;
  }
}


src/exchangeGroups.js
import _ from 'lodash';

const exchangeDefaults = {
  type: 'direct',
  durable: true,
  autoDelete: false,
};

const queueDefaults = {
  autoDelete: false,
  durable: true,
  noAck: false,
};

function groupFromInput(param, { name, persistent }) {
  let returnVal = {};
  if (_.isString(param)) {
    returnVal = { name: param };
  } else if (_.isObject(param)) {
    returnVal = param;
  }
  returnVal.name = returnVal.name || name;
  returnVal.persistent = returnVal.persistent || persistent;
  return returnVal;
}

function normalizeExchangeGroup(key, group) {
  if (!group.keys) {
    throw new Error('Must specify keys for each group');
  }

  const normalized = {
    readLimit: group.readLimit || 100,
    retryDelay: group.retryDelay || 10000,
    retries: _.isNumber(group.retries) ? group.retries : 5,
    keys: group.keys,
  };
  const persistent = group.persistent || false;

  normalized.exchange = groupFromInput(group.exchange, { name: key, persistent });
  const exchangeName = normalized.exchange.name;

  normalized.queue = groupFromInput(group.queue, { name: `${exchangeName}.q` });

  if (normalized.retries) {
    normalized.retryExchange = groupFromInput(group.retryExchange, { name: `${exchangeName}.retry`, persistent });
    normalized.retryQueue = groupFromInput(group.retryQueue, { name: `${normalized.retryExchange.name}.q` });
    const ttlKey = group.perMessageTtl === true ? 'perMessageTtl' : 'messageTtl';
    normalized.retryQueue[ttlKey] = normalized.retryDelay;
  }

  if (normalized.retries || group.rejectedExchange || group.rejectedQueue) {
    normalized.rejectedExchange = groupFromInput(group.rejectedExchange, { name: `${exchangeName}.rejected`, persistent });
    normalized.rejectedQueue = groupFromInput(group.rejectedQueue, { name: `${normalized.rejectedExchange.name}.q` });
  }

  return normalized;
}

export function normalizeExchangeGroups(exchangeGroups) {
  const normalized = {};

  Object.entries(exchangeGroups).forEach(([k, v]) => {
    normalized[k] = normalizeExchangeGroup(k, v);
  });

  return normalized;
}

function addBinding(fooFooConfig, exchange, queue, keys) {
  fooFooConfig.exchanges.push(exchange);
  fooFooConfig.queues.push(queue);
  fooFooConfig.bindings.push({
    exchange: exchange.name,
    target: queue.name,
    keys,
  });
}

export function fooFooConfigFromExchangeGroups(exchangeGroups) {
  const foofooConfig = {
    exchanges: [],
    queues: [],
    bindings: [],
  };

  Object.entries(exchangeGroups).forEach(([, g]) => {
    const exchange = Object.assign({}, exchangeDefaults, g.exchange);
    const queue = Object.assign({}, queueDefaults, g.queue);
    queue.limit = g.readLimit;

    if (g.retryExchange) {
      const retryExchange = Object.assign({}, exchangeDefaults, g.retryExchange);
      const retryQueue = Object.assign({}, queueDefaults, g.retryQueue);
      retryQueue.deadLetter = exchange.name;
      addBinding(foofooConfig, retryExchange, retryQueue, g.keys);
    }

    if (g.rejectedExchange) {
      const rejectedExchange = Object.assign({}, exchangeDefaults, g.rejectedExchange);
      const rejectedQueue = Object.assign({}, queueDefaults, g.rejectedQueue);

      addBinding(foofooConfig, rejectedExchange, rejectedQueue, g.keys);

      queue.deadLetter = rejectedExchange.name;
    }

    addBinding(foofooConfig, exchange, queue, g.keys);
  });

  return foofooConfig;
}


src/parseShorthandConfig.js
import _ from 'lodash';

const exchangeDefaults = {
  type: 'direct',
  durable: true,
  autoDelete: false,
};

const queueDefaults = {
  autoDelete: false,
  durable: true,
  noAck: false,
};

export default function (shorthandConfigs) {
  const output = {
    exchanges: [],
    queues: [],
    bindings: [],
  };

  function addBinding(exchange, queue, keys) {
    output.exchanges.push(exchange);
    output.queues.push(queue);
    output.bindings.push({
      exchange: exchange.name,
      target: queue.name,
      keys:keys,
    });
  }

  shorthandConfigs.forEach((c) => {
    const keys = c.keys;
    if (!keys) {
      throw new Error('Must specify keys');
    }

    let exchange = _.isString(c.exchange) ? { name: c.exchange } : c.exchange;
    if (!exchange.name) {
      throw new Error('Exchanges must be named.');
    }
    exchange = Object.assign({}, exchangeDefaults, exchange);

    let queue = (_.isString(c.queue) ? { name: c.queue } : c.queue) || {};
    if (!queue.name) {
      throw new Error('Queues must be named.');
    }
    queue = Object.assign({}, queueDefaults, queue);

    if (!c.noRetry) {
      const retryExchange = Object.assign({}, exchange);
      retryExchange.name = `${retryExchange.name}.retry`;

      const retryQueue = Object.assign({}, queue);
      retryQueue.name = `${retryQueue.name}.retry`;
      retryQueue.messageTtl = 250;
      retryQueue.deadLetter = exchange.name;

      addBinding(retryExchange, retryQueue, keys);

      const rejectedExchange = Object.assign({}, exchange);
      rejectedExchange.name = `${exchange.name}.rejected`;

      const rejectedQueue = Object.assign({}, queue);
      rejectedQueue.name = `${queue.name}.rejected`;

      addBinding(rejectedExchange, rejectedQueue, keys);

      queue.deadLetter = rejectedExchange.name;
    }

    addBinding(exchange, queue, keys);
  });
  return output;
}


src/index.js
import { EventEmitter } from 'events';
import assert from 'assert';
import foofoo from 'foo-foo-mq';
import _ from 'lodash';
import {
  normalizeExchangeGroups,
  fooFooConfigFromExchangeGroups,
} from './exchangeGroups';
import { WrappedMessage } from './WrappedMessage';
import boleLogger from './boleLogger';

const exchangeErrorRE = /Failed to create exchange '(.*)' on connection 'default'/;
const ORIGINAL_ARGS = Symbol('Original rabbitmq options');

function isDev() {
  return process.env.NODE_ENV === 'development' || !process.env.NODE_ENV;
}

async function delay(ms) {
  return new Promise(accept => setTimeout(accept, ms));
}

function finalConfigFromConfig(context, opts, mqConnectionConfig) {
  let exchangeGroups = (opts.config && opts.config.exchangeGroups) || {};
  exchangeGroups = normalizeExchangeGroups(exchangeGroups);

  const exchangeGroupConfig = fooFooConfigFromExchangeGroups(exchangeGroups);
  const finalConfig = Object.assign({}, opts.config);
  delete finalConfig.exchangeGroups;

  finalConfig.exchanges = (finalConfig.exchanges || []).concat(exchangeGroupConfig.exchanges);
  finalConfig.queues = (finalConfig.queues || []).concat(exchangeGroupConfig.queues);
  finalConfig.bindings = (finalConfig.bindings || []).concat(exchangeGroupConfig.bindings);
  finalConfig.connection = Object.assign({}, finalConfig.connection, mqConnectionConfig);

  let dependencies = (opts.config && opts.config.dependencies) || [];

  const dependencyAttribs = {};
  // Create dependencies in test mode.
  if (process.env.NODE_ENV !== 'test') {
    dependencyAttribs.passive = true;
  }
  dependencies = _.map(dependencies, d => Object.assign({ name: d }, dependencyAttribs));

  finalConfig.exchanges = finalConfig.exchanges.concat(dependencies);

  if (opts.logging) {
    foofoo.log(boleLogger(context.logger, Array.isArray(opts.logging) ? opts.logging : undefined));
  }

  finalConfig.exchangeGroups = exchangeGroups;
  return finalConfig;
}

export default class FooFooClient extends EventEmitter {
  constructor(context, opts) {
    super();
    assert(opts, 'configured-rabbitmq-client must be passed arguments');
    assert(opts.username, 'configured-rabbitmq-client missing username setting');
    assert(opts.password, 'configured-rabbitmq-client missing password setting');

    const hostname = Array.isArray(opts.hostname) ? opts.hostname.join(',') : opts.hostname;

    context.logger.info('Initializing RabbitMQ client', {
      user: opts.username,
      host: hostname || 'rabbitmq',
    });
    const mqConnectionConfig = {
      // In our usage, we've found that a short timeout causes trouble
      // on startup with spurious connection errors just because the box
      // is busy. So we default to a higher initial timeout
      timeout: 15000,
      // We don't typically use reply queues and they tend to cruft up
      // the server, so disable by default
      replyQueue: false,
      ...opts.connectionOptions,
      user: opts.username,
      pass: opts.password,
      host: hostname || 'rabbitmq',
      port: opts.port || 5672,
      vhost: opts.basePath || '/',
    };

    // Event handlers need to be cleaned up afterwards...
    this.connSubscription = foofoo.on('connected', () => {
      context.logger.info('RabbitMQ connection established.');
    });

    this[ORIGINAL_ARGS] = { opts, mqConnectionConfig };
    this.finalConfig = finalConfigFromConfig(context, opts, mqConnectionConfig);
    this.exchangeGroups = this.finalConfig.exchangeGroups;
    delete this.finalConfig.exchangeGroups;
    this.originalContext = context;
    this.contextFunction = opts.contextFunction;
    this.startedCalled = false;
    this.subs = [];
    this.client = foofoo;
    this.subscriptions = opts.subscriptions;
  }

  // Function should recieve a queue message and return a gb-services style context.

  publish(...args) {
    return this.client.publish(...args);
  }

  request(...args) {
    return this.client.request(...args);
  }

  bulkPublish(...args) {
    return this.client.bulkPublish(...args);
  }

  async start(context) {
    assert(!this.startCalled, 'start called multiple times on configured-rabbitmq-client instance');
    this.startCalled = true;

    const maxRetries = 5;
    for (let retries = maxRetries; retries >= 0; retries -= 1) {
      try {
        context.logger.info('Configuring RabbitMQ client');
        // eslint-disable-next-line no-await-in-loop
        await foofoo.configure(this.finalConfig);
        break;
      } catch (stringError) {
        if (process.env.MQ_MAKE_EXCHANGES && exchangeErrorRE.test(stringError.message)) {
          const [, failedQueue] = stringError.message.match(exchangeErrorRE);
          const { opts, mqConnectionConfig } = this[ORIGINAL_ARGS];
          _.pull(opts.config.dependencies, failedQueue);
          opts.config.exchangeGroups = opts.config.exchangeGroups || {};
          opts.config.exchangeGroups[failedQueue] = { keys: '*' };
          this.finalConfig = finalConfigFromConfig(context, opts, mqConnectionConfig);
          this.exchangeGroups = this.finalConfig.exchangeGroups;
          delete this.finalConfig.exchangeGroups;
          retries += 1; // this one doesn't count

          // Unforunately, foofoomq seems to need time to sort its sh** out.
          context.logger.info('Queue setup failed. Waiting 2.5s then creating queue', { name: failedQueue });
          // eslint-disable-next-line no-await-in-loop
          await delay(2500);
        }
        if (retries) {
          // This comparison basically is only false when we've done the auto-create thing above
          if (retries <= maxRetries) {
            context.logger.warn(`Queue configuration failed, retrying ${retries} more times`, stringError);
            if (isDev() && exchangeErrorRE.test(stringError)) {
              context.logger.info(`
*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*

Exchanges are missing. Typically this comes from dependent services which you are not
running. In order to create these queues with default characteristics, set
the MQ_MAKE_EXCHANGES environment variable and restart.

*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*`);
            }
          }

          try {
            // eslint-disable-next-line no-await-in-loop
            await foofoo.shutdown();
            foofoo.reset();
            context.logger.info('RabbitMQ cleanup complete');
          } catch (inner) {
            context.logger.warn('RabbitMQ cleanup failed', {
              error: inner.message,
              stack: inner.stack,
            });
          }

          if (retries <= maxRetries) {
            // eslint-disable-next-line no-await-in-loop
            await delay((1 + (maxRetries - retries)) * 2000);
          }
        } else {
          if (typeof stringError === 'string') {
            throw new Error(stringError);
          }
          throw stringError;
        }
      }
    }

    foofoo.nackUnhandled();
    foofoo.nackOnError();

    if (typeof this.subscriptions === 'object') {
      for (const [, sub] of Object.entries(this.subscriptions)) {
        this.subscribe(sub.queue, sub.type, sub.handler);
      }
    }

    this.closeSubscription = foofoo.on('closed', () => {
      if (!this.shuttingDown) {
        context.logger.error('RabbitMQ connection was closed.');
      }
    });
    this.unreachSubscription = foofoo.on('unreachable', () => {
      // TODO shutdown the process?
      context.logger.error('RabbitMQ connection has failed permanently.');
      this.emit('unreachable');
    });
    this.failSubscription = foofoo.onReturned('failed', (e) => {
      context.logger.error('RabbitMQ connection has failed.', {
        error: e.message,
        stack: e.stack,
      });
    });

    return this;
  }

  async stop(context) {
    assert(this.startCalled, 'stop called multiple times on configured-rabbitmq-client instance');
    context.logger.info('Closing RabbitMQ connection');
    await Promise.all(this.subs.map((s) => {
      s[0].remove();
      return FooFooClient.gracefulQueueShutdown(s[1]);
    }));
    this.shuttingDown = true;
    this.connSubscription.unsubscribe();
    delete this.connSubscription;
    if (this.closeSubscription) {
      this.closeSubscription.unsubscribe();
      delete this.closeSubscription;
    }
    if (this.unreachSubscription) {
      this.unreachSubscription.unsubscribe();
      delete this.unreachSubscription;
    }
    await foofoo.shutdown();
    foofoo.reset();
    this.startCalled = false;
  }

  async subscribe(exchangeGroupOrQueue, type, handler) {
    if (process.env.DISABLE_RABBITMQ_SUBSCRIPTIONS === 'true') {
      return;
    }
    if (process.env.DISABLE_RABBITMQ_SUBSCRIPTIONS) {
      const disabled = process.env.DISABLE_RABBITMQ_SUBSCRIPTIONS.split(',');
      if (disabled.includes(exchangeGroupOrQueue) || disabled.includes(`${exchangeGroupOrQueue}$${type}`)) {
        return;
      }
    }

    let wrappedHandler = async (rmqMessage) => {
      const message = new WrappedMessage(this, rmqMessage);
      if (handler.length === 2) {
        const context = this.contextFunction
          && await this.contextFunction(this.originalContext, message);
        await handler(context, message);
      } else {
        await handler(message);
      }
    };

    let finalQueueName = exchangeGroupOrQueue;
    const exchangeGroup = this.exchangeGroups[exchangeGroupOrQueue];

    if (exchangeGroup) {
      finalQueueName = exchangeGroup.queue.name;
      if (exchangeGroup.retries) {
        wrappedHandler = async (rmqMessage) => {
          const message = new WrappedMessage(this, rmqMessage);
          let context;
          try {
            if (handler.length === 2) {
              context = this.contextFunction
                && await this.contextFunction(this.originalContext, message);
              await handler(context, message);
            } else {
              await handler(message);
            }
          } catch (e) {
            const headers = message.properties.headers || {};
            const retryCount = (headers.retryCount || 0) + 1;
            headers.retryCount = retryCount;
            headers.error = e.message;

            // Error logging
            if (context) {
              const loggingMetadata = context.gb.wrapError(e);
              loggingMetadata.queueName = finalQueueName;
              loggingMetadata.type = message.type;
              loggingMetadata.routingKey = message.fields.routingKey;
              loggingMetadata.retryCount = retryCount;
              context.gb.logger.error('Error handling queue message.', loggingMetadata);
            }

            if (retryCount > exchangeGroup.retries || e.noRetry) {
              message.reject();
            } else {
              const messageOptions = {
                type: message.type,
                body: message.body,
                routingKey: message.fields.routingKey,
                correlationId: message.properties.correlationId,
                timestamp: message.properties.timestamp,
                expiresAfter: exchangeGroup.retryQueue.perMessageTtl || undefined,
                headers,
              };
              await this.publish(exchangeGroup.retryExchange.name, messageOptions);
              message.ack();
            }
          }
        };
      }
    }

    const handlerThunk = foofoo.handle(type, wrappedHandler, finalQueueName);
    let mq = foofoo.getQueue(finalQueueName);
    if (!mq) {
      // Let's give it a try.
      mq = foofoo.getQueue(`${finalQueueName}.q`);
    }
    assert(mq, `Specified queue (${finalQueueName}) does not exist`);
    mq.subscribe(false);
    this.subs.push([handlerThunk, mq]);
  }

  // eslint-disable-next-line class-methods-use-this
  metadata() {
    return {
      activeMessages: Array.from(WrappedMessage.activeMessages()),
    };
  }

  static async gracefulQueueShutdown(q) {
    if (q?.lastQueue?.messages?.messages?.length) {
      return new Promise((accept) => {
        q.lastQueue.messages
          .on('empty', accept)
          .once();
      });
    }
    // Mostly for tests which restart right away, but Rabbit is finicky
    return delay(1000);
  }

  static get activeMessages() {
    return WrappedMessage.activeMessages();
  }
}

export class MockRabbitClient {
  constructor() {
    this.subscriptions = {};
    this.publishMocks = {};
  }

  async internalPublish(exchange, keyOrOptions, maybeBody) {
    let key = keyOrOptions;
    let body = maybeBody;
    if (typeof key !== 'string') {
      key = keyOrOptions.type;
      ({ body } = keyOrOptions);
    }
    const mock = (this.publishMocks[exchange] || {})[key];
    if (mock) {
      await mock(exchange, key, body);
    } else {
      this.context.logger.info(`Unmocked messaged published. Exchange: ${exchange} Key: ${key}`);
    }
  }

  async publish(...args) {
    return this.internalPublish(...args);
  }

  async request(...args) {
    return this.internalPublish(...args);
  }

  async bulkPublish(bulk) {
    if (Array.isArray(bulk)) {
      return Promise.all(
        bulk.map(({ type, body, exchange }) => this.internalPublish(exchange, type, body)),
      );
    }
    return Promise.all(
      Object.entries(bulk)
        .map(([exchange, { type, body }]) => this.internalPublish(exchange, type, body)),
    );
  }

  async start(context) {
    this.context = context;
    return this;
  }

  async subscribe(queueName, key, handler) {
    this.subscriptions[queueName] = this.subscriptions[queueName] || {};
    this.subscriptions[queueName][key] = handler;
  }

  resetPublishMocks() {
    this.publishMocks = {};
  }

  async mockPublish(exchange, key, handler) {
    this.publishMocks[exchange] = this.publishMocks[exchange] || {};
    this.publishMocks[exchange][key] = handler;
  }

  async testMessage(context, queueName, key, body, {
    ack = () => { },
    nack = () => { },
    reject = () => { },
  } = {}) {
    const handler = (this.subscriptions[queueName] || {})[key];
    if (!handler) {
      throw new Error(`No handler for key ${key} on queue ${queueName}`);
    } else {
      const message = {
        body,
        ack,
        nack,
        reject,
      };
      if (handler.length === 2) {
        await handler(context, message);
      } else {
        await handler(message);
      }
    }
  }
}
